src/lib.rs

```diff
@@ -295,7 +295,7 @@ impl InstructionOperands {
                         polarity: o0 >> 2,
                         source,
                         index: o1,
-                        relative: o1 & 0b10000 != 0,
+                        relative: (o1 & 0b10000 != 0) && source == WaitSource::IRQ,
```

Comment:
It would be more consistent to return `None` for an invalid combination, instead of just ignoring the wrongly set relative flag in the binary encoding.

Reply:
If the source is not IRQ, the relative argument is irrelevant so the more idiomatic representation would be to have relative be an argument of `WaitSource::IRQ(bool)`. While decoding, there's no inconsistency.

This fix is to avoid miss interpretation (and a panic) when doing a decode-encode sequence (like what `Pio::install` does).

Reply:
When `rel`  is set it follows the same rules as `IRQ` but its still very confusing to me.
It looks like the address is always 3 bits (which would make sense because there are 7 possible irq) so the actual irq selected seems to be:


```rust
let rel = (o1 & 0b10000) != 0;
let addr = o1 & 0b00111;

let irq = (addr & 0b100) | (if rel { (addr + (SMI::id() as u8)) } else { addr } & 0b011)
```

Reply:
I meant something else:
When `o1` is `0b10000`, the assembly should be `wait ... irq 0 rel`, ie. `InstructionOperands::WAIT { index: 0, rel: true, ... }`
But the current code will set `index` to `16` → `InstructionOperands::Wait { index: 16, rel: true, ... }`

Reply:
Yes, the index should probably be `o1 & 0b111` but I'm not 100% sure.

Reply:
Also, some of the test cases actually use invalid values:
```
instr_test!(wait(0, WaitSource::IRQ, 10, false), 0b001_00000_010_01010);
instr_test!(wait(1, WaitSource::IRQ, 15, false), 0b001_00000_110_01111);
```
But the last few comments are only tangential. While there may still be bugs, those are old and not really related to this pull request.